File: render_book.py
import os
import shutil
import subprocess
import sys

def run_command(command):
    """Menjalankan perintah terminal dan menampilkan outputnya secara realtime."""
    process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
    while True:
        output = process.stdout.readline()
        if output == '' and process.poll() is not None:
            break
        if output:
            print(output.strip())
    rc = process.poll()
    return rc

def clean_dir(path):
    """Menghapus direktori jika ada."""
    if os.path.exists(path):
        if os.path.isdir(path):
            shutil.rmtree(path)
        else:
            os.remove(path)

def swap_freeze(target_freeze):
    """Mengganti folder _freeze utama dengan cache khusus format."""
    clean_dir("_freeze")
    if os.path.exists(target_freeze):
        shutil.copytree(target_freeze, "_freeze")
        print(f"-> Menggunakan cache dari '{target_freeze}'")
    else:
        print(f"-> Tidak ada cache sebelumnya untuk '{target_freeze}'. Memulai cache baru.")

def save_freeze(target_freeze):
    """Menyimpan cache dari _freeze ke folder khusus format."""
    clean_dir(target_freeze)
    if os.path.exists("_freeze"):
        shutil.copytree("_freeze", target_freeze)
        print(f"-> Menyimpan cache ke '{target_freeze}'")
    clean_dir("_freeze")

def render_html():
    print("\n=== MERENDER FORMAT HTML ===")
    swap_freeze("html_freeze")
    
    # Render ke html
    exit_code = run_command("quarto render --to html")
    if exit_code != 0:
        print("Error saat merender HTML!")
        sys.exit(exit_code)
        
    # Pindahkan output ke folder khusus html agar tidak tertumpuk format lain
    dest_dir = "_book_html"
    clean_dir(dest_dir)
    if os.path.exists("_book"):
        shutil.move("_book", dest_dir)
        print(f"-> Output HTML berhasil disimpan di '{dest_dir}'")
        
    save_freeze("html_freeze")

def render_pdf():
    print("\n=== MERENDER FORMAT PDF (TYPST) ===")
    swap_freeze("typst_freeze")
    
    # Render ke typst (menghasilkan file .pdf)
    exit_code = run_command("quarto render --to typst")
    if exit_code != 0:
        print("Error saat merender PDF/Typst!")
        sys.exit(exit_code)
        
    # Pindahkan file pdf hasil render ke direktori root utama
    if os.path.exists("_book"):
        for file in os.listdir("_book"):
            if file.endswith(".pdf"):
                src_path = os.path.join("_book", file)
                shutil.copy2(src_path, file)
                print(f"-> File PDF berhasil disalin ke root: {file}")
        shutil.rmtree("_book")
        
    save_freeze("typst_freeze")

if __name__ == "__main__":
    # Jalankan render untuk kedua format
    render_html()
    render_pdf()
    print("\nProses render selesai dengan sukses!")

File: render_book.R
# render_book.R
# Skrip untuk merender buku ke berbagai format tanpa konflik cache

render_html <- function() {
  cat("Memulai render format HTML...\n")
  tmpdir <- tempfile()
  dir.create(tmpdir)
  
  # Copy file ke tempdir
  fs::dir_copy(".", tmpdir)
  
  # Bersihkan _freeze jika terbawa
  if (fs::dir_exists(file.path(tmpdir, "_freeze"))) {
    fs::dir_delete(file.path(tmpdir, "_freeze"))
  }
  
  # Gunakan cache khusus html
  if (fs::dir_exists(file.path(tmpdir, "html_freeze"))) {
    fs::dir_copy(file.path(tmpdir, "html_freeze"), file.path(tmpdir, "_freeze"))
  }
  
  # Eksekusi render
  quarto::quarto_render(tmpdir, output_format = "html", as_job = FALSE)
  
  # Pindahkan _book yang dihasilkan
  if (fs::dir_exists("_book_html")) fs::dir_delete("_book_html")
  fs::file_move(file.path(tmpdir, "_book"), "_book_html")
  
  # Simpan kembali cache
  if (fs::dir_exists("html_freeze")) fs::dir_delete("html_freeze")
  if (fs::dir_exists(file.path(tmpdir, "_freeze"))) {
    fs::file_move(file.path(tmpdir, "_freeze"), "html_freeze")
  }
  
  cat("Selesai render HTML. Hasil ada di folder '_book_html'.\n\n")
}

render_typst <- function() {
  cat("Memulai render format Typst...\n")
  tmpdir <- tempfile()
  dir.create(tmpdir)
  
  fs::dir_copy(".", tmpdir)
  
  if (fs::dir_exists(file.path(tmpdir, "_freeze"))) {
    fs::dir_delete(file.path(tmpdir, "_freeze"))
  }
  
  if (fs::dir_exists(file.path(tmpdir, "typst_freeze"))) {
    fs::dir_copy(file.path(tmpdir, "typst_freeze"), file.path(tmpdir, "_freeze"))
  }
  
  quarto::quarto_render(tmpdir, output_format = "typst", as_job = FALSE)
  
  # Cari file PDF hasil Typst
  pdf_files <- fs::dir_ls(file.path(tmpdir, "_book"), glob = "*.pdf")
  if (length(pdf_files) > 0) {
    # Pindahkan PDF ke root direktori kerja utama
    fs::file_copy(pdf_files[1], basename(pdf_files[1]), overwrite = TRUE)
    cat("File PDF berhasil disalin:", basename(pdf_files[1]), "\n")
  } else {
    cat("Peringatan: File PDF hasil render Typst tidak ditemukan.\n")
  }
  
  if (fs::dir_exists("typst_freeze")) fs::dir_delete("typst_freeze")
  if (fs::dir_exists(file.path(tmpdir, "_freeze"))) {
    fs::file_move(file.path(tmpdir, "_freeze"), "typst_freeze")
  }
  
  cat("Selesai render Typst.\n\n")
}

render_all <- function() {
  render_html()
  render_typst()
  cat("Semua format (HTML dan Typst) berhasil dirender.\n")
}
